src/frontend/src/optimizer/rule/top_n_on_index_rule.rs

```diff
@@ -0,0 +1,133 @@
+//  Copyright 2023 RisingWave Labs
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+// This source code is licensed under both the GPLv2 (found in the
+// COPYING file in the root directory) and Apache 2.0 License
+// (found in the LICENSE.Apache file in the root directory).
+
+use risingwave_common::util::sort_util::OrderType;
+
+use super::{BoxedRule, Rule};
+use crate::optimizer::plan_node::{LogicalLimit, LogicalScan, LogicalTopN, PlanTreeNodeUnary};
+use crate::optimizer::property::{Direction, FieldOrder, Order};
+use crate::optimizer::PlanRef;
+
+pub struct TopNOnIndexRule {}
+
+impl Rule for TopNOnIndexRule {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let logical_top_n: &LogicalTopN = plan.as_logical_top_n()?;
+        let logical_scan: LogicalScan = logical_top_n.input().as_logical_scan()?.to_owned();
+        let order = logical_top_n.topn_order();
+        if order.field_order.is_empty() {
+            return None;
+        }
+        if let Some(p) = self.try_on_pk(logical_top_n, logical_scan.clone(), order) {
+            Some(p)
+        } else {
+            self.try_on_index(logical_top_n, logical_scan, order)
+        }
+    }
+}
+
+impl TopNOnIndexRule {
+    pub fn create() -> BoxedRule {
+        Box::new(TopNOnIndexRule {})
+    }
+
+    fn try_on_index(
+        &self,
+        logical_top_n: &LogicalTopN,
+        logical_scan: LogicalScan,
+        order: &Order,
+    ) -> Option<PlanRef> {
+        let index = logical_scan.indexes().iter().find(|idx| {
+            Order {
+                field_order: idx
+                    .index_table
+                    .pk()
+                    .iter()
+                    .map(|idx_item| FieldOrder {
+                        index: idx_item.index,
+                        direct: idx_item.direct,
+                    })
+                    .collect(),
+            }
+            .satisfies(order)
```

Comment:
We can't compare index order directly with the topn order, because they refer to different columns. We need a map between them. The following case can reproduce this bug.
```sql
create table t (a int primary key, b int); 
create index idx on t(b);
explain  select * from t order by b limit 1;
```

Reply:
Then we should test if the index order is a _superset_ of the topn order?

Reply:
We need it and actually `satisfies` method has already handled the superset problem, but considering the above case, my point is we need to compare index order with topn order at the same level, that is, both refer to primary table column or index column.

Reply:
got it, I'll try to solve it in the next commit